### ISSUE TITLE
Derived table ETL tools

### ISSUE THREAD

**Issue by karlcz** [nih-cfde/cfde-deriva#47]
I think we want to produce a couple of composable tool (or a single tool with multiple sub-functions) to help generate derived data tables which will assist the portal and dashboards.  These tools will work with a scratch database to allow the use of SQL, but in the end will produce datapackages in the same format as those expected from the DCCs (a frictionless table-schema connected to TSV files containing row data).

1. Create and initialize an empty database as a workspace
2. Load a datapackage into database
   - Translate frictionless table definitions into SQL DDL and create (empty) tables
   - Insert row data from TSV files in frictionless file `path` references
3. Generate derived tables based on custom ETL datapackage
   - Translate frictionless tablel definitions into SQL DDL and create (empty) tables
   - Insert row data produced by SQL query in frictionless file `derivation_sql_path` references
4. Dump derived tables as datapackage
   - Reuse custom ETL datapackage as output frictionless schema template?
   - Prune custom SQL ETL query references
   - Dump tables to TSV and add `path` references to frictionless schema

The general case would involve repeating the step (2) to load more than one datapackage and then compute the total results over the whole set.

Then, the resulting derived table datapackage should be valid to send through our normal loader code, so that the set of input DCC datapackages and the derived datapackage can all be loaded into a single catalog.

I would like us to assume/require PostgreSQL for the scratch database, so that we know which SQL features are available.

I committed a [draft ETL input](https://github.com/nih-cfde/cfde-deriva/blob/load_202006/table-schema/demo-202006-denorm.json) which defines a few tables and provides supporting SQL files for each.  I think this sort of structure will be good for maintenance, i.e. a single manifest file that defines the tables structurally and provides the domain-specific logic to compute the result.

**Comment by rpwagner** [comment]
@jgaff I'm wondering if this can be added to the existing action (maybe with a flag to toggle it) where the new datapackage gets created and stored on the endpoint before trying to load it into DERIVA using the existing loading code. That way a user can download the transformed bag if something fails.

**Comment by karlcz** [comment]
On May 27, Rick Wagner modulated:
> @jgaff I'm wondering if this can be added to the existing action (maybe
> with a flag to toggle it) where the new datapackage gets created and
> stored on the endpoint before trying to load it into DERIVA using the
> existing loading code. That way a user can download the transformed bag
> if something fails.
>

To be clear, I also think this tool needs to exist in library+CLI form
to operate against a local DB in a dev environment, no submission to
a hosted environment being required/assumed.

But I would expect that we would want to include it in the submission
service flow too.

**Comment by rpwagner** [comment]
I agree, it should be like the other package that currently does the load into DERIVA. Something that can be used in a few contexts. But this transformation should be a hidden part of the submission process for DCC users.

**Comment by jgaff** [comment]
@rpwagner I see no issues with adding this tool to the ingest Action Provider. If the (successful, non-error) output from the tool is always valid to ingest into Deriva, we may not need a flag for the user to worry about, though. And if the transformation fails/returns an error, I think the Action should attempt to ingest the user's data without modification (similarly to how `.apply_custom_config()` is currently run, but exceptions/errors are not treated as ingest failures).

**Comment by karlcz** [comment]
Right, the output should be valid unless we are debugging the ETL
definitions themselves.  The DCC-oriented submitter is probably not
the same person who would be pursuing those sorts of bugs, and the one
doing that work is more likely to want to use the ETL tool locally,
i.e. to prepare a local sandbox DB and examine it with ad hoc SQL
queries, rather than just inspect TSV files.

But, it is probably important to make the partial-load status apparent
to the user somehow?  It will affect the subsequent GUI presentation
of their data, and we wouldn't want them getting frustrated with
misunderstood and hidden failures...  (this applies equally to the
existing custom config step and the new ETL steps).

**Comment by karlcz** [comment]
We should probably look at https://github.com/frictionlessdata/tableschema-sql-py  to help with the initial steps to load the DCC datapackage into postgres.  The getting started documentation seems pretty close to the goal already, except that it is configuring a sqlite connection instead of postgres...

**Comment by carlkesselman** [comment]
Yes.

I know this is not in your requirements… but is having a deriva-py based version of this something that would be useful as well?

Carl

----------------------------------------------------------

Dr. Carl Kesselman
Dean’s Professor, Epstein Department of Industrial and Systems Engineering
Fellow, Information Sciences Institute Viterbi School of Engineering Professor,

Preventive Medicine Keck School of Medicine

University of Southern California
4676 Admiralty Way, Suite 1001, Marina del Rey, CA  90292-6695
Phone:  +1 (310) 448-9338
 Email:  carl@isi.edu
Web: http://www.isi.edu/~carl
On May 28, 2020, 11:23 AM -0700, Karl Czajkowski <notifications@github.com>, wrote:

We should probably look at https://github.com/frictionlessdata/tableschema-sql-py to help with the initial steps to load the DCC datapackage into postgres. The getting started documentation seems pretty close to the goal already, except that it is configuring a sqlite connection instead of postgres...

**Comment by karlcz** [comment]
What I think you are describing would really be a different tableschema-deriva-py driver that provides a similar python API to load/store datapackages from an ermrest catalog.
While that might have utility for some other use cases, it isn't relevant to this issue to build a tool for our goal of performing the SQL-based ETL (well actually LTE, but that sure seems like a confusing acronym).
